### ISSUE TITLE
what happend to the sqlite results?

### ISSUE THREAD

**Issue by timm** [ai-se/HOW1#26]
fig8 has sqlite results but https://github.com/ai-se/HOW1/issues/24 does not. 

say what?

t

**Comment by rahlk** [comment]
Must have missed that, I'll add it right away.

**Comment by timm** [comment]
fyi: my schedule is work on the paper till noon, get out into the sun,
back at 2pm and working on from there.

what's your schedule? what exams /class projects do you have left to complete?

**Comment by rahlk** [comment]
I have a paper due tomorrow and a final at noon on Wednesday.

**Comment by timm** [comment]
how much time do you need for those? i.e. realistically, without messing up your GPA, how much time can you devote sunday... wednesday to increasing "G" or recall scores for these data sets?

**Comment by rahlk** [comment]
I can try spending about 10 hours towards this. BTW have you seen my latest
comment on issue #24 https://github.com/ai-se/HOW1/issues/24?

**Comment by timm** [comment]
Thats not a comment. that's data. and you left it without comment. so i had to do the interpretation. which i did. half a hour to produce the following. showing that very few of our data sets had pd > 0.6 and pf < 0.4. suggesting  that you need to do some work to convince reviewers of the merits of this technique for improving defect prediction

data. initial pd pf
ant        a 0.54 0.13
camel      c 0.45 0.38
ivy        i 0.8  0.41
jedit      j 0.64 0.3
log4j      l 0.8  0.19
luncene    L 0,6  0.3
pbeans(rf) P 0.9  0.32 
poi        p 0.57 0.33
synapse    s 0.63 0.49
velocity   v 0.86 0.66
xalan      x 0.44 0
xerces     X 0.22 0.4

![image](https://cloud.githubusercontent.com/assets/29195/7446520/526750e2-f1a9-11e4-9af6-c10168748e3a.png)

i dream of the day when my students don;t tell me what is, they tell me what it is about and what they are going to do about it.

you can:
- pull off algorithm magic showing that some more entries in the pd > 06 and pf < 0.4
- do a quick lit review on these data sets to show that the numbers we are generating are better/equal to most 
- do something else.

so, phd student at ncstate, what is your plan?

**Comment by rahlk** [comment]
Sorry about that, I'll ensure this won't happen again.

Only other algorithmic magic I can think of is to use DE to tune (but my
studies have shown that it's not very helpful. The pd/pf scores saturate at
what we saw). I can try tuning smote's sub/super sampling limits to see if
there is a _magic_ ratio that improves the prediction scores.

I'll also do a lit. survey to see if people have worked on this before.

Best,
_Rahul_
_rkrsn.us_ http://rkrsn.us

​

Rahul Krishna
_​​__The RAISE Lab http://ai4se.net/_
_Research Assistant​,_
_Electrical Engineering,_
_North Carolina State University._

**Comment by timm** [comment]
if you just focus on precision, wei reports that this can be increased dramatically by DEs. and, as you recall, DEs were not slow (1000 evals is way too many; he got away with 50 to 90 evals)

**Comment by timm** [comment]
i really want the defect results but if that fucks up, could just go with the configuration stuff. anything there more than berkly db, sqltive VVLM,? 286?

**Comment by rahlk** [comment]
They have only 6 datasets.

​

Rahul Krishna
_​​__The RAISE Lab http://ai4se.net/_
_Research Assistant​,_
_Electrical Engineering,_
_North Carolina State University._

**Comment by rahlk** [comment]
I'm afraid we might have too little to write if use only the SVEN dataset.
Also, If we have only 6 defect datasets, that should not stop us from not
reporting the results at all. Right?

​

Rahul Krishna
_​​__The RAISE Lab http://ai4se.net/_
_Research Assistant​,_
_Electrical Engineering,_
_North Carolina State University._

**Comment by timm** [comment]
Six defect data sets, culled from12, is good. shows we can't plan for everything. But I really want six with pd > 0.6 and pf < 0.4. or forgetting pd,pf, precision > 70 or even 80

**Comment by rahlk** [comment]
I see. I'll try my best to get these results. I believe the class imbalance
in the test data will play a major role. Since we can't do anything to
handle that, I am not sure how much improvement there will be.

​

Rahul Krishna
_​​__The RAISE Lab http://ai4se.net/_
_Research Assistant​,_
_Electrical Engineering,_
_North Carolina State University._

**Comment by rahlk** [comment]
I think I've fixed it, I can give you 5 Datasets (Ant, Ivy, Jedit, Lucene,
Poi) that have pf>0.6 and pf<0.4 and one with pd = 0.59 and pf = 0.38
(Camel).

This is what I did---
- Use versions 1 to N-1 to train and N to test.
- SMOTE the training data
- Tune for precision with DE (keeping track of the best set of values and
SMOTE while tuning as well)

Here are my results:



​

Rahul Krishna
*​​**The RAISE Lab <http://ai4se.net/>*
*Research Assistant​,*
*Electrical Engineering,*
*North Carolina State University.*

**Comment by rahlk** [comment]
I think I've fixed it, I can give you 5 Datasets (Ant, Ivy, Jedit, Lucene,
Poi) that have pf>0.6 and pf<0.4 and one with pd = 0.59 and pf = 0.38
(Camel).

This is what I did---
- Use versions 1 to N-1 to train and N to test.
- SMOTE the training data
- Tune for precision with DE (keeping track of the best set of values and
  SMOTE while tuning as well)

Here are my results:

![stats](https://cloud.githubusercontent.com/assets/1433964/7459676/a2d721e8-f26b-11e4-9b80-2f67a325b5b5.png)

**Comment by timm** [comment]
cool.  four questions"

1) you tuned for precision? then what are the precision values? 

2)  did you get the same results as Wei? i.e. DE terminated VERY early (say, 90 evals).

3)  can you do the above graph in gnuplot and use pd, pf for the axis (and include source code for same into the repo)

4) can i get a table saying something line

| data set | train | test |
| --- | --- | --- |
| ant | 456 classes from 1.2,1.3 | 125 classes from 1.4 |
| ivy | 670  classes from 1.2,1.3 | 130 classes from 1.4 |

t

**Comment by rahlk** [comment]
- DE did terminate early except for 2 datasets, log4j and velocity (both of them had very bad pd pf)
- When you say classes, do you mean all classes, or only the defective classes?

**Comment by timm** [comment]
all classes.

very interested in the precision scores BEFORE and AFTER de-ing

**Comment by rahlk** [comment]
I accidentally cleared my screen, let me run the rig again to produce the precision scores.

**Comment by rahlk** [comment]
```
## ant       


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.63 | 0.82 | 0.51      | 0.78     | 0.57 | 0.72 |
| RF(SMOTE)      | 0.66 | 0.79 | 0.47      | 0.76     | 0.55 | 0.72 |
+----------------+------+------+-----------+----------+------+------+

## camel     


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.62 | 0.59 | 0.27      | 0.60     | 0.38 | 0.61 |
| RF(SMOTE)      | 0.50 | 0.70 | 0.28      | 0.66     | 0.36 | 0.58 |
+----------------+------+------+-----------+----------+------+------+

## ivy       


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.65 | 0.75 | 0.26      | 0.75     | 0.37 | 0.71 |
| RF(SMOTE)      | 0.65 | 0.65 | 0.19      | 0.65     | 0.30 | 0.65 |
+----------------+------+------+-----------+----------+------+------+

## jedit     


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.64 | 0.72 | 0.05      | 0.72     | 0.09 | 0.67 |
| RF(SMOTE)      | 0.50 | 0.73 | 0.04      | 0.72     | 0.07 | 0.59 |
+----------------+------+------+-----------+----------+------+------+

## log4j     


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.30 | 0.91 | 0.97      | 0.35     | 0.46 | 0.46 |
| RF(SMOTE)      | 0.32 | 0.94 | 0.98      | 0.37     | 0.49 | 0.48 |
+----------------+------+------+-----------+----------+------+------+

## lucene    


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.60 | 0.73 | 0.77      | 0.65     | 0.68 | 0.66 |
| RF(SMOTE)      | 0.55 | 0.75 | 0.77      | 0.62     | 0.63 | 0.63 |
+----------------+------+------+-----------+----------+------+------+

## poi       


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.61 | 0.88 | 0.90      | 0.71     | 0.73 | 0.72 |
| RF(SMOTE)      | 0.51 | 0.68 | 0.73      | 0.56     | 0.60 | 0.57 |
+----------------+------+------+-----------+----------+------+------+

## synapse   


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.47 | 0.85 | 0.61      | 0.72     | 0.53 | 0.61 |
| RF(SMOTE)      | 0.45 | 0.82 | 0.56      | 0.70     | 0.50 | 0.58 |
+----------------+------+------+-----------+----------+------+------+

## velocity  


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.77 | 0.39 | 0.39      | 0.52     | 0.52 | 0.51 |
| RF(SMOTE)      | 0.79 | 0.39 | 0.40      | 0.53     | 0.53 | 0.53 |
+----------------+------+------+-----------+----------+------+------+

## xalan     


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.48 | 1.00 | 1.00      | 0.49     | 0.65 | 0.65 |
| RF(SMOTE)      | 0.41 | 0.95 | 1.00      | 0.41     | 0.58 | 0.56 |
+----------------+------+------+-----------+----------+------+------+

## xerces    


+----------------+------+------+-----------+----------+------+------+
| Dataset        | Pd   | 1-Pf | Precision | Accuracy | F    | G    |
+----------------+------+------+-----------+----------+------+------+
| RF(SMOTE+tune) | 0.39 | 0.67 | 0.77      | 0.46     | 0.52 | 0.49 |
| RF(SMOTE)      | 0.30 | 0.72 | 0.76      | 0.41     | 0.43 | 0.42 |
+----------------+------+------+-----------+----------+------+------+
```

**Comment by timm** [comment]
looking..

**Comment by timm** [comment]
darn. better precision does not optimize for better pd,pf. oh well

**Comment by timm** [comment]
what params were u tuning for RF?

**Comment by rahlk** [comment]
Yeah, I think I know why...

See, we have many cases with fewer defects than non defects, therefore, even a few extra False Positives in the denominator of prec = TP/TP+FP will greatly diminish the precision. However, in the grand scheme of things those few extra FP's do not affect pd's and pf's

**Comment by timm** [comment]
did u tune for the same range of aprams as wei in the random forests?

**Comment by rahlk** [comment]
Yup, I literally took the ranges form the tunelearners paper to do my tunings. I played with SMOTE and incorporated that into tuning. I also rewrote DE to keep track of the best feature set found (since we are only doing a single objective optimization.)

**Comment by rahlk** [comment]
I am running another test with a binary domination on pd, pf.

**Comment by timm** [comment]
ok- unless you see errors in the stuff below or unless you have other tricks to try (*), please close this issue

(*) if you have other tricks. please update deltas1.tex in the repo (it describes this table)

so here's what i got from your results.

i think i can rationalize the use of ant,ivy,jedit,lucense,poi for this study. please change the text and the tables of data accordingly.

![screenshot 2015-05-04 20 44 12](https://cloud.githubusercontent.com/assets/29195/7465458/6a277e1e-f29e-11e4-916e-f2d80a2688a9.png)

somethings to note here
- looking at the percentiles. pd is rarely hurt but often helped a lot by tuning
- looking at the percentiles, pf is hurt as much as it is harmed so nothing to say here
- when tuning helps, it REALLY helps , see the *\* and **\* examples.

so, in summary, not all data sets support planning and when some don't a little tuning can really fix things up

**Comment by rahlk** [comment]
Oh sure and CAMEL too, it's pd is 60 and pf 41.

**Comment by timm** [comment]
nope. if our bordeline is 60,40 then not camel
